src/vm/syncblk.cpp

```diff
@@ -1931,22 +1935,46 @@ AwareLock::EnterHelperResult ObjHeader::EnterObjMonitorHelperSpin(Thread* pCurTh
                 return AwareLock::EnterHelperResult_UseSlowPath;
             }
 
-            // Check the recursive case once before the spin loop. If it's not the recursive case in the beginning, it will not
-            // be in the future, so the spin loop can avoid checking the recursive case.
             SyncBlock *syncBlock = g_pSyncTable[oldValue & MASK_SYNCBLOCKINDEX].m_SyncBlock;
             _ASSERTE(syncBlock != NULL);
             AwareLock *awareLock = &syncBlock->m_Monitor;
-            if (awareLock->EnterHelper(pCurThread, true /* checkRecursiveCase */))
+
+            AwareLock::EnterHelperResult result = awareLock->TryEnterBeforeSpinLoopHelper(pCurThread);
```

Comment:
I realize that what I am about to comment on is not part of the delta in this PR, but I am trying to rationalize the whole code flow here.    My mental model for the EnterObjMonitorHelperSpin helper is that it was there to avoid making an awareLock (otherwise even ONE contention on a lock at any time would force the awareLock to be created).   However we have an awareLock at this point, so the question is why can't we just return UseSlowPath and avoid this code?    This code *should* be a clone of what is done in AwareLock, and the extra cost of getting there (making an awareLock (but we have one) and setting up a transition frame (but we need to spin anyway), should not be an issue.     It would be nice to simply remove this code.   What prevents this?

Reply:
I was thinking about that, it's kind of a pity to have to force-inline the moderately slow-path 
spinning code (which it used to do in the asm helpers to avoid calls). The overhead of the transition is not something I fully understand but my high level guess is that taking the slow path is equivalent in overhead to a p/invoke, which is fairly high. I'm not sure if it will matter though, probably worth a test. Moving spinning to the slow path would also significantly improve register allocation in the fast path since it removes the last offending call. I plan on looking into this more closely as part of https://github.com/dotnet/coreclr/issues/14571.

Yes avoiding creating the aware lock is another advantage of spinning more (when it's a thin lock anyway). With this change a monitor would transition into an aware lock much more easily, there are perf/behavior/memory tradeoffs involved and I think this is a decent one.

Spinning can't be avoided altogether once we have an aware lock. Spinning just a little bit significantly improves perf under contention and avoids excessive context switching in scenarios involving smaller time scales, which I'm sure are very common.

Regarding moving the spin path to the slow path, I don't think it matters whether we have an an aware lock already or not, it's the same kind of situation for a thin lock.

Reply:
Just FYI,  Yes, a HelperMethodFrame transition is like a PINVOKE, but it is slower (because it does something called epilog walking to set up the frame).   I have not measured it recently, but a normal optimized PINVOKE is 20-30 clock ticks.   HelperMethodFrames are probably 2-3 times that.     There is some goo to get from there to the AwareLock.Enter() where we again try to take the lock with in interlocked operation.    But it seems like that overhead is basically on the same order as what we would spin for, so we should not be too concerned about the overhead. 

As you say it has the benefit of making the register allocation leaner for the paths that remain.   But I also care about the fact that we are cloning non-trivial code (we have a very similar spin wait loop in EnterEpilogHelper).   The less cloned code we have the better (for example we don't do the ETW contention logging in TryEnterObjMonitorSpinHelper, and arguably we should.   It would also be SO easy for the logic that counts spinners to be off in one place but not the other).   

I agree we need spinning in the aware-lock case.  I just want it done one place if possible, and it seems like we are doing it in EnterEpilogHelper so that should be enough.    

The value of trying this now, is that we don't have to deal with any bugs associated with the code we are about to remove.   I suppose it could wait until #14571, but if you see no issue, and removing it does not cause regressions it is a very nice improvement in my opinion...
 
On the issue of spinning to avoid creating an awareLock:   I would not spin more just to avoid Aware lock creation.   If spinning just a bit more is what saved you, odds are that sooner or later you will 'fall off' that edge and need to make the lock.     The key part to handle the case where this is little to no contention (which is common in many cases).

Reply:
I take back some of what I said above.    I see that the logic in EnterEpilogHelper does the Wait before is does any spinning (it does the spinning on waking up).    So you can't just remove the code in EnterEpilogHelper  without losing spinning in the AwareLock case before waiting.

I must admit I would prefer that it work the way I thought it worked (that there is no logic in   EnterObjMonitorHelperSpin to AwareLock (we always bail to the slow helper)).   This would require that AwareLock.EnterEpilog to spin before waiting (it is the same loop, you just spin before you wait).    Thus all AwareLock logic is now centralized (LockState actually does no need to be in a header-file, it is only used in AwareLock's code in SyncBlk.cpp).   

But that is not a trivial change (not huge, but not trivial), and would need to be perf-tested.    I will not ask you to make that change to get this fix check in.  

Please do consider it, however.  I think it would simplify the code and as you mention, may help #14571.

Reply:
For what it is worth, it seems that EnterObjMonitorHelper EnterObjMonitorHelperSpin are only called from one place TryEnterObjMonitorSpinHelper, and there are some tricky conditions that are cloned in each of these methods.    You could effectively combine them to EnterObjMonitorHelperSpin which pretty much just EnterObjMonitorHelper with a loop that calls SpinWait 

This of course only works well if we first kick all the AwareLock stuff out of  EnterObjMonitorHelperSpin suggested above.  

While I like this idea, I am not suggesting you do it, just giving you food for thought.

Reply:
RE HelperMethodFrame overhead, the overhead may even be worth it as it would add a potentially beneficial delay before the first try in the spin loop. Can't be sure until I try it.

> The value of trying this now, is that we don't have to deal with any bugs associated with the code we are about to remove. I suppose it could wait until #14571, but if you see no issue, and removing it does not cause regressions it is a very nice improvement in my opinion...

I would like to keep this effort separate from this PR due to time constraints. My estimation is that this is not a short task. If it would be beneficial in avoiding bugs, we could consider postponing this PR but in that event a smaller PR that fixes the YieldProcessor issue for Skylake processors would be necessary in the interim in case the full set of changes don't make it into 2.1.

> But I also care about the fact that we are cloning non-trivial code (we have a very similar spin wait loop in EnterEpilogHelper). The less cloned code we have the better
> It would also be SO easy for the logic that counts spinners to be off in one place but not the other

The spin loops in EnterObjMonitorHelperSpin and EnterEpilogHelper are distinct and most of their logic is not shared:
- EnterObjMonitorHelperSpin spin loop for AwareLock
  - Before spin loop, atomically attempt to acquire lock or register spinner (must not register spinner if lock is available)
  - In spin loop, if lock is available, atomically acquire lock and unregister spinner
  - After spin loop, unregister spinner, but if lock was available at the time of unregistering the spinner, the lock must be acquired and the spinner unregistered atomically unless the lock is stolen by another thread, the comment there explains why
- EnterEpilogHelper spin loop
  - Precondition: waiter is registered
  - In spin loop, if lock is available, atomically acquire lock and clear waiter-signaled-to-wake bit
  - After spin loop, clear waiter-signaled-to-wake bit, but if lock was available at the time of clearing the bit, the lock must be acquired and the waiter unregistered atomically unless the lock is stolen by another thread. If lock is not acquired, it will proceed to waiting again.

There are also differences relevant to mitigating waiter starvation. There is some similarly in the flow and it's likely possible to refactor the flow to be shared, but to me it doesn't seem worth it.

> we don't do the ETW contention logging in TryEnterObjMonitorSpinHelper, and arguably we should

This would be far too verbose. Considering how short the spin duration is, from the perspective of the event I think it's fair to consider that there is no contention until waiting is necessary. It may even already be too verbose compared to before, but it's something that can be fixed later if it turns out to be an issue.

> For what it is worth, it seems that EnterObjMonitorHelper EnterObjMonitorHelperSpin are only called from one place TryEnterObjMonitorSpinHelper, and there are some tricky conditions that are cloned in each of these methods.

Agreed, and there's certainly more refactoring that can be done to simplify this code. At this point though, I think a separate PR would be appropriate for those kinds of changes.

> You could effectively combine them to EnterObjMonitorHelperSpin which pretty much just EnterObjMonitorHelper with a loop that calls SpinWait

This is actually what I tried initially but this showed significant perf regressions on Linux and I decided to keep them separate and not inline the spin loop. On Windows there was not much difference in perf. Since we're thinking about moving the spinning to the slow helper path anyway, it may make sense to keep them separate at least until that idea can be tested.

Reply:
Filed issue https://github.com/dotnet/coreclr/issues/15095 to track monitor code cleanup